crates/rpc/src/cairo/starknet_rs/estimate.rs

```diff
@@ -0,0 +1,133 @@
+    execution_state: ExecutionState,
+    message: FunctionCall,
+    sender_address: EthereumAddress,
+) -> Result<FeeEstimate, CallError> {
+    // prepend sender address to calldata
+    let sender_address =
+        Felt::from_be_slice(sender_address.0.as_bytes()).expect("Ethereum address is 160 bits");
+    let calldata = std::iter::once(pathfinder_common::CallParam(sender_address))
+        .chain(message.calldata.into_iter())
+        .map(|p| p.0.into())
+        .collect();
+
+    let transaction = L1Handler::new(
+        starknet_in_rust::utils::Address(message.contract_address.get().into()),
+        message.entry_point_selector.0.into(),
+        calldata,
+        0.into(),
+        execution_state.chain_id.0.into(),
+        None,
+    )?;
+    let transaction = Transaction::L1Handler(transaction);
+
+    let mut result = estimate_fee_impl(execution_state, vec![transaction])?;
+
+    if result.len() != 1 {
+        return Err(
+            anyhow::anyhow!("Internal error: expected exactly one fee estimation result").into(),
+        );
+    }
+
+    let result = result.pop().unwrap();
+
+    Ok(result)
+}
+
+fn estimate_fee_impl(
+    execution_state: ExecutionState,
+    transactions: Vec<Transaction>,
+) -> Result<Vec<FeeEstimate>, CallError> {
+    let gas_price = execution_state.gas_price;
+    let block_number = execution_state.block_number;
+
+    let (mut state, block_context) = execution_state.into_state()?;
+
+    let mut fees = Vec::with_capacity(transactions.len());
+    for (transaction_idx, transaction) in transactions.iter().enumerate() {
+        let span = tracing::debug_span!("execute", transaction_hash=%super::transaction::transaction_hash(transaction), %block_number, %transaction_idx);
+        let _enter = span.enter();
+
+        let transaction_for_simulation = transaction.create_for_simulation(false, false, true);
+        let tx_info = transaction_for_simulation.execute(&mut state, &block_context, 1_000_000);
+
+        match tx_info {
+            Ok(tx_info) => {
+                if let Some(revert_error) = tx_info.revert_error {
+                    tracing::info!(%revert_error, "Transaction reverted");
+                    return Err(CallError::Reverted(revert_error));
+                }
+
+                tracing::trace!(actual_fee=%tx_info.actual_fee, "Transaction estimation finished");
+                // L1Handler transactions don't normally calculate the fee -- we have to do that after execution.
+                let actual_fee = match transaction {
+                    Transaction::L1Handler(_) => calculate_tx_fee(
+                        &tx_info.actual_resources,
+                        gas_price.as_u128(),
+                        &block_context,
+                    )?,
+                    _ => tx_info.actual_fee,
+                };
+
+                fees.push(FeeEstimate {
+                    gas_consumed: U256::from(actual_fee) / std::cmp::max(1.into(), gas_price),
```

Comment:
Is the minimum only there in-case gas price is zero?
```suggestion
                    gas_consumed: U256::from(actual_fee) / gas_price.min(1.into()),
```